### ISSUE TITLE
Nondeterministic output of processors

### ISSUE THREAD

**Issue by jglick** [stapler/stapler#219]
https://github.com/stapler/stapler/blob/3765ae49d6ff355ab457927eeb5be5776b54817b/core/src/main/java/org/kohsuke/stapler/jsr269/AbstractProcessorImpl.java#L66-L70 can result in build products from a given source tree including a timestamp, breaking artifact diff tools.

It would also be better for `*.properties`-formatted outputs to be sorted, for maximum stability.

**Comment by jglick** [comment]
Can probably just `trim()` values, and then reimplement

> For each entry the key string is written, then an ASCII =, then the associated element string. For the key, all space characters are written with a preceding \ character. For the element, leading space characters, but not embedded or trailing space characters, are written with a preceding \ character. The key and element characters #, !, =, and : are written with a preceding backslash to ensure that they are properly loaded.
> Characters not in Latin-1 in the comments are written as \uxxxx for their appropriate unicode hexadecimal value xxxx.
> Characters less than \u0020 and characters greater than \u007E in property keys or values are written as \uxxxx for the appropriate hexadecimal value xxxx.

**Comment by daniel-beck** [comment]
This specification cannot be complete, it doesn't mention that `\` needs to get a preceding `\` as well.

I'm tempted to write to memory and then just dump the first (comment) line before writing to a file.

**Comment by jglick** [comment]
> dump the first (comment) line

IIRC you can pass an explicit header which overrides the autogenerated date header. You would indeed still need to sort explicitly.

> assuming it never breaks lines for length

Not that I know of, but better check.

**Comment by daniel-beck** [comment]
> IIRC you can pass an explicit header which overrides the autogenerated date header. You would indeed still need to sort explicitly.

https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#store-java.io.Writer-java.lang.String- documents that the autogenerated comment is always included.

**Comment by jglick** [comment]
Indeed. AFAICT you can just `| egrep -v '^#' | sort` and have it work even for exotic keys and values. (In our cases, the keys are always going to something like Java identifiers so that is probably not a concern, but values can sometimes be multiline strings with whatever characters, right?)
